File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
def check():
    for i in range(4):
        if acgt[i] > cnt[i]:
            return 0
    return 1

sLen, pLen = map(int, input().split())
S = input()
idx = { 'A': 0, 'C': 1, 'G': 2, 'T': 3 }
acgt = [*map(int, input().split())]
cnt = [0, 0, 0, 0]
for i in range(pLen):
    cnt[idx[S[i]]] += 1
ans = check()
for i in range(pLen, sLen):
    cnt[idx[S[i]]] += 1
    cnt[idx[S[i - pLen]]] -= 1
    ans += check()
print(ans)

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.java
import java.io.*;
import java.util.*;

public class Main {

	static int[] acgt = new int[4], acgtCnt = new int[4];
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int sLen = Integer.parseInt(st.nextToken());
		int pLen = Integer.parseInt(st.nextToken());
		String s = br.readLine();
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < 4; i++) {
			acgt[i] = Integer.parseInt(st.nextToken());
		}
		
		for (int i = 0; i < pLen; i++) {
			++acgtCnt[idx(s.charAt(i))];
		}
		int ans = check();
		for (int i = pLen; i < sLen; i++) {
			--acgtCnt[idx(s.charAt(i - pLen))];
			++acgtCnt[idx(s.charAt(i))];
			ans += check();
		}
		System.out.println(ans);
	}
	
	public static int check() {
		for (int i = 0; i < 4; i++) {
			if (acgt[i] > acgtCnt[i])
				return 0;
		}
		return 1;
	}
	
	public static int idx(char v) {
		switch (v) {
		case 'A':
			return 0;
		case 'C':
			return 1;
		case 'G':
			return 2;
		case 'T':
			return 3;
		}
		return 0;
	}
	
}